src/main/java/org/spongepowered/api/command/parameters/specification/ValueParameterModifiers.java

```diff
@@ -0,0 +1,160 @@
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * Class containing {@link ValueParameterModifier}s.
+ */
+public final class ValueParameterModifiers {
+
+    private ValueParameterModifiers() {}
+
+    private static final ValueParameterModifierFactory factory = DummyObjectProvider.createFor(ValueParameterModifierFactory.class, "factory");
+
+    /**
+     * Indicates that the parameter should continue to parse all remaining
+     * elements in the provided set of arguments.
+     */
+    public static ValueParameterModifier allOf() {
+        return CatalogedValueParameterModifiers.ALL_OF;
+    }
+
+    /**
+     * Indicates that there should only be one element returned by the
+     * parameter.
+     */
+    public final ValueParameterModifier onlyOne() {
+        return CatalogedValueParameterModifiers.ONLY_ONE;
+    }
+
+    /**
+     * Indicates that the parameter is optional, but will throw an exception
+     * if an argument exists to be parsed.
+     */
+    public final ValueParameterModifier optional() {
+        return CatalogedValueParameterModifiers.OPTIONAL;
+    }
+
+    /**
+     * Indicates that the parameter is optional, and will not prevent the rest
+     * of the command arguments from being parsed if this cannot parse an
+     * argument.
+     */
+    public final ValueParameterModifier optionalWeak() {
+        return CatalogedValueParameterModifiers.OPTIONAL_WEAK;
+    }
+
+    /**
+     * Requires the parameter to be provided a certain number of times.
+     *
+     * <p>Command values will be stored under their provided keys in the
+     * {@link CommandExecutionContext}.</p>
+     *
+     * @param times The number of times to repeat the element.
+     * @return The {@link ValueParameterModifier}
+     */
+    public static ValueParameterModifier repeated(int times) {
+        return factory.repeated(times);
+    }
+
+    /**
+     * If after a {@link ValueParameter}'s execution there is no value stored
+     * in the associated key, this modifier will put the supplied default into
+     * the context.
+     *
+     * @param defaultValue The default value.
+     * @return The {@link ValueParameterModifier}
+     */
+    public static ValueParameterModifier defaultValue(Object defaultValue) {
+        return factory.defaultValue(defaultValue);
+    }
+
+    /**
+     * If after a {@link ValueParameter}'s execution there is no value stored
+     * in the associated key, this modifier will put the default obtained from
+     * the supplied {@link Supplier} into the context, if the default is
+     * non-null.
+     *
+     * @param defaultValueSupplier The default value {@link Supplier}.
+     * @return The {@link ValueParameterModifier}
+     */
+    public static ValueParameterModifier defaultValueSupplier(Supplier<Object> defaultValueSupplier) {
```

Comment:
It would be awesome if this had an overload that accepted `Function<CommandSource, Object>`. Example use-case would be custom versions of `fooOrSource()`.